src/Shared/PlatformNegotiation.cs

```diff
@@ -0,0 +1,110 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Evaluation;
+
+namespace Microsoft.Build.Shared
+{
+    /// <summary>
+    /// This class contains only static methods, which are used in both the 
+    /// tasks and graph projects in order for two projects to negotiate which platform a projectreference
+    /// should be built as.
+    /// </summary>
+    static internal class PlatformNegotiation
+    {
+        internal static string GetNearestPlatform(string referencedProjectPlatform, string projectReferencePlatformsMetadata, string projectReferenceLookupTableMetadata, string platformLookupTable, string projectPath, string currentProjectPlatform, TaskLoggingHelper? log = null)
+        {
+            Dictionary<string, string>? currentProjectLookupTable = ExtractLookupTable(platformLookupTable, log);
```

Comment:
This looks pretty allocatey. In a big graph, it'll be parsing + allocating a new dictionary per project, right? Can that be avoided, especially since I expect there will only be a few distinct lookup tables?

(This doesn't have to block this PR but let's make sure we're thinking about it.)

Reply:
Noting this for a future PR. I assume well refactor some of this logic eventually once the logic for dual inner and outer build properties are allowed